File: 0907-koko-eating-bananas/0907-koko-eating-bananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int maxi=INT_MIN;
        int ans = INT_MAX;
        for(int n:piles){
            maxi=max(maxi,n);
        }
        int low=1, high=maxi;
        while(low<high){
            int mid=low+(high-low)/2;
            if( hoursToEat(piles, mid,h) <= h){
                high=mid;
            }else{
                low=mid+1;
            }
        }
        return high;
    }
    int hoursToEat(vector<int>& piles,int mid_val,int hours){
        int totalHoursNeeded = 0;
        for(int pile : piles){
            if(pile % mid_val !=0)
                totalHoursNeeded += (pile/mid_val + 1);
            else 
                totalHoursNeeded += (pile/mid_val);
        }
        return totalHoursNeeded;
    }
};

File: 0907-koko-eating-bananas/0907-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int max= Arrays.stream(piles).max().getAsInt();
        int min = 1;
        while( min < max){
            int mid = min+(max-min)/2;
            if(check(piles, mid, h)){
                max = mid;
            }else{
                min=mid+1;
            }
        }
        return min;
    }
    public boolean check(int[] piles, int mid, int h){
        int total=0;
        for(int pile:piles){
            if(pile%mid!=0)
            total +=1;
            total += pile/mid;
        }
        return total <= h;
    }
}